apps/emqx/src/emqx_broker_helper.erl

```diff
@@ -194,6 +195,45 @@ code_change(_OldVsn, State, _Extra) ->
 %% Internal functions
 %%--------------------------------------------------------------------
 
+collect_and_handle(Reg0, Down0) ->
+    {Reg, Down} = collect_messages(Reg0, Down0),
+    %% handle register before handle down to avoid race condition
+    %% because down message is always the last one from a process
+    ok = handle_register(Reg),
+    ok = handle_down(Down).
+
+collect_messages(Reg, Down) ->
+    collect_messages(Reg, Down, ?BATCH_SIZE).
+
+%% Collect both register_sub and 'DOWN' messages in a loop.
+%% There is no other message sent to this process, so the
+%% 'receive' should not have to scan the mailbox.
+collect_messages(Reg, Down, 0) ->
+    {Reg, Down};
+collect_messages(Reg, Down, N) ->
+    receive
+        {register_sub, Pid, Id} ->
+            collect_messages([{Id, Pid} | Reg], Down, N - 1);
+        {'DOWN', _MRef, process, Pid, _Reason} ->
+            collect_messages(Reg, [Pid | Down], N - 1)
+    after 0 ->
+        {Reg, Down}
+    end.
+
+handle_register(Reg) ->
+    lists:foreach(fun do_handle_register/1, Reg).
+
+do_handle_register({SubId, SubPid}) ->
+    true = (SubId =:= undefined) orelse ets:insert(?SUBID, {SubId, SubPid}),
+    _ = erlang:monitor(process, SubPid),
```

Comment:
Q: Can we simplify further, and just rely on a single monitor set up by `emqx_cm`? Say, make it submit a composite job to the generic pool.

Reply:
that'd be the next PR.

Reply:
https://github.com/emqx/emqx/pull/16413